tls/s2n_cipher_suites.c

```diff
@@ -1074,6 +1074,33 @@ int s2n_cipher_suites_cleanup(void)
     return 0;
 }
 
+S2N_RESULT s2n_iana_to_cipher_suite(const uint8_t iana[], struct s2n_cipher_suite **cipher_suite)
+{
+    ENSURE_REF(cipher_suite);
+    *cipher_suite = NULL;
+    ENSURE_REF(iana);
+
+    int low = 0;
+    int top = s2n_array_len(s2n_all_cipher_suites) - 1;
+
+    /* Perform a textbook binary search */
+    while (low <= top) {
+        /* Check in the middle */
+        size_t mid = low + ((top - low) / 2);
+        int m = memcmp(s2n_all_cipher_suites[mid]->iana_value, iana, S2N_TLS_CIPHER_SUITE_LEN);
+
+        if (m == 0) {
+            *cipher_suite = s2n_all_cipher_suites[mid];
+            return S2N_RESULT_OK;
+        } else if (m > 0) {
+            top = mid - 1;
+        } else if (m < 0) {
+            low = mid + 1;
+        }
+    }
+    return S2N_RESULT_OK;
```

Comment:
Should this return `S2N_RESULT_OK`? If we get here haven't we failed to find the suite?

Reply:
I'm not sure. We've failed to find the suite, but we indicate that by setting the output cipher suite as NULL.

I think it comes down to the behavior we want:
* An unsupported cipher suite is an error, and if the method succeeds then cipher suite is guaranteed not to be NULL. If the method fails, you don't know whether the cipher suite isn't supported / doesn't exist or whether something went wrong in the search.
* An unsupported cipher suite is fine, and an error indicates something went wrong in the search. The caller checks whether cipher suite is NULL to determine whether the cipher suite is unsupported.

I can see an argument for either approach, and I don't think S2N has defined any consistent behavior around this :\

Reply:
I'm inclined towards "Success means we found something". We don't currently have any valid uses for calling this where a `NULL` would be a good response. The closes we come is a test where we can check that the failure code is appropriate. I think that this behavior is safer as it means that there is no longer a second null check for people to risk forgetting.

Reply:
> An unsupported cipher suite is an error, and if the method succeeds then cipher suite is guaranteed not to be NULL. If the method fails, you don't know whether the cipher suite isn't supported / doesn't exist or whether something went wrong in the search.

Being able to distinguish between reasons could potentially be useful (which would support for keeping as is and adding documentation around the contract with the caller).